numpy/core/src/multiarray/array_coercion.c

```diff
@@ -0,0 +1,1434 @@
+ *
+ * 1. The user calls ``np.array(..., dtype=np.dtype("<f8"))``
+ * 2. The user calls ``np.array(..., dtype="S")``
+ * 3. The user calls ``np.array(...)``
+ *
+ * In the first case, in principle only the shape needs to be found. In the
+ * second case, the DType class (e.g. string) is already known but the DType
+ * instance (e.g. length of the string) has to be found.
+ * In the last case the DType class needs to be found as well. Note that
+ * it is not necessary to find the DType class of the entire array, but
+ * the DType class needs to be found for each element before the actual
+ * dtype instance can be found.
+ *
+ * Further, there are a few other things to keep in mind when coercing arrays:
+ *
+ *   * For UFunc promotion, Python scalars need to be handled specially to
+ *     allow value based casting.  This requires python complex/float to
+ *     have their own DTypes.
+ *   * It is necessary to decide whether or not a sequence is an element.
+ *     For example tuples are considered elements for structured dtypes, but
+ *     otherwise are considered sequences.
+ *     This means that if a dtype is given (either as a class or instance),
+ *     it can effect the dimension discovery part.
+ *     For the "special" NumPy types structured void and "c" (single character)
+ *     this is special cased.  For future user-types, this is currently
+ *     handled by providing calling an `is_known_scalar` method.  This method
+ *     currently ensures that Python numerical types are handled quickly.
+ *
+ * In the initial version of this implementation, it is assumed that dtype
+ * discovery can be implemented sufficiently fast.  That is, it is not
+ * necessary to create fast paths that only find the correct shape e.g. when
+ * ``dtype=np.dtype("f8")`` is given.
+ *
+ * The code here avoid multiple conversion of array-like objects (including
+ * sequences). These objects are cached after conversion, which will require
+ * additional memory, but can drastically speed up coercion from from array
+ * like objects.
+ */
+
+
+/*
+ * For finding a DType quickly from a type, it is easiest to have a
+ * a mapping of pytype -> DType.
+ * TODO: This mapping means that it is currently impossible to delete a
+ *       pair of pytype <-> DType.  To resolve this, it is necessary to
+ *       weakly reference the pytype. As long as the pytype is alive, we
+ *       want to be able to use `np.array([pytype()])`.
+ *       It should be possible to retrofit this without too much trouble
+ *       (all type objects support weak references).
+ */
+PyObject *_global_pytype_to_type_dict = NULL;
+
+
+/* Enum to track or signal some things during dtype and shape discovery */
+enum _dtype_discovery_flags {
+    FOUND_RAGGED_ARRAY = 1,
+    GAVE_SUBCLASS_WARNING = 2,
+    PROMOTION_FAILED = 4,
+    DISCOVER_STRINGS_AS_SEQUENCES = 8,
+    DISCOVER_TUPLES_AS_ELEMENTS = 16,
+    MAX_DIMS_WAS_REACHED = 32,
+    DESCRIPTOR_WAS_SET = 64,
```

Comment:
Nit: Can we write these in the form `1 << x`, so people don't have to memorize powers of two?